lib/AST/DiagnosticConsumer.cpp

```diff
@@ -14,15 +14,148 @@
+
+    Optional<unsigned> bufferID = SM.getIDForBufferIdentifier(pair.first);
+    assert(bufferID.hasValue() && "consumer registered for unknown file");
+    CharSourceRange range = SM.getRangeForBuffer(bufferID.getValue());
+    LocationToConsumerMap.emplace_back(range, i);
+  }
+
+  // Sort the "map" by buffer /end/ location, for use with std::lower_bound
+  // later. (Sorting by start location would produce the same sort, since the
+  // ranges must not be overlapping, but since we need to check end locations
+  // later it's consistent to sort by that here.)
+  std::sort(LocationToConsumerMap.begin(), LocationToConsumerMap.end(),
+            [](const MapEntry &left, const MapEntry &right) -> bool {
+    auto compare = std::less<const char *>();
+    return compare(getRawLoc(left.first.getEnd()).getPointer(),
+                   getRawLoc(right.first.getEnd()).getPointer());
+  });
+
+  // Check that the ranges are non-overlapping.
+  assert(LocationToConsumerMap.end() ==
+           std::adjacent_find(LocationToConsumerMap.begin(),
+                              LocationToConsumerMap.end(),
+                              [](const MapEntry &left, const MapEntry &right) {
+                                return left.first.overlaps(right.first);
+                              }) &&
+         "overlapping ranges despite having distinct files");
+}
+
+DiagnosticConsumer *
+FileSpecificDiagnosticConsumer::consumerForLocation(SourceManager &SM,
+                                                    SourceLoc loc) const {
+  if (loc.isInvalid())
+    return nullptr;
+
+  // This map is generated on first use and cached, to allow the
+  // FileSpecificDiagnosticConsumer to be set up before the source files are
+  // actually loaded.
+  if (LocationToConsumerMap.empty()) {
+    auto *mutableThis = const_cast<FileSpecificDiagnosticConsumer*>(this);
+    mutableThis->buildLocationToConsumerMap(SM);
+  }
+
+  // "Find the first range whose end location is greater than or equal to
+  // 'loc'."
+  auto possiblyContainingRangeIter =
+      std::lower_bound(LocationToConsumerMap.begin(),
+                       LocationToConsumerMap.end(),
+                       loc,
+                       [](const std::pair<CharSourceRange, unsigned> &entry,
+                          SourceLoc loc) -> bool {
+    auto compare = std::less<const char *>();
+    return compare(getRawLoc(entry.first.getEnd()).getPointer(),
+                   getRawLoc(loc).getPointer());
+  });
+
+  if (possiblyContainingRangeIter != LocationToConsumerMap.end() &&
+      possiblyContainingRangeIter->first.contains(loc)) {
+    return SubConsumers[possiblyContainingRangeIter->second].second.get();
+  }
+
+  return nullptr;
+}
+
+void FileSpecificDiagnosticConsumer::handleDiagnostic(
+    SourceManager &SM, SourceLoc Loc, DiagnosticKind Kind,
+    StringRef FormatString, ArrayRef<DiagnosticArgument> FormatArgs,
+    const DiagnosticInfo &Info) {
+
+  DiagnosticConsumer *specificConsumer;
```

Comment:
I would factor the switch out into a function that returned the specificConsumer (or null). The `specificConsumer` variable becomes local to the first three cases that way, so it's less code to trace through in a reader's head.

Reply:
I still don't think that's worth it, especially since there's an update to `ConsumerForSubsequentNotes`. Lots of small functions are *also* hard to reason about.